### DISCUSSION TITLE
My blank app does not work properly with elevated privilege.

### DISCUSSION THREAD

**Discussion by lansyin** [microsoft/WindowsAppSDK#671] [Q&A] [answered]
> I apologize for my lame English in advance. I’m new to windows app development, so perhaps I’m doing stuff wrong. Sorry. 

I created a new `Blank App, Packaged (WinUI 3 in Desktop)` C# project. Then added `<rescap:Capability Name="allowElevation" />` to `Package.appmanifest`. I found this app displays nothing when I run it as administrator (`MainWindow` is expected). 

Though I did not add `requireAdministrator` to `app.manifest`, it still crashed when I right click the app icon and select "Run as Administrator". 

The same problem exists in my C++/WinRT project. I don't know whether a WinUI 3 app can work with elevated privilege. There are something I found before requesting your help.

- [debugging-elevated-winui-in-desktop](https://developercommunity.visualstudio.com/t/debugging-elevated-winui-in-desktop/1291065) 
- [Run as administrator crashes WinUI Desktop App](https://github.com/microsoft/microsoft-ui-xaml/issues/3046)

Software versions are listed below:

- IDE:
  - Microsoft Visual Studio Community 2019 Preview Version 16.10.0 Preview 1.0
- Extensions:
  - Project Reunion Version 0.5.0.85112000
  - C++/WinRT Version 2.0.210329.4
- NuGet Packages:
  - Microsoft.ProjectReunion/0.5.0
  - Microsoft.ProjectReunion.Foundation/0.5.0
  - Microsoft.ProjectReunion.WinUI/0.5.0

**Comment by andrewleader** [comment, marked as answer]
Hi @catsalwaysmeow, sorry for the delay. Unfortunately, **WinUI 3 currently does not support running elevated**. This is due to a limitation in WinRT in the OS (WinRT activation doesn't work in elevated apps), and we'll hopefully be fixing that in a future version of Windows.

**If you need elevation support, vote and tell us here: https://portal.productboard.com/winappsdk/1-windows-app-sdk/c/37-elevated-admin-support**

**Reply by wjk** [reply]
Is this a fix that can be backported as part of Reunion, or do we have to wait until the relevant Windows 10 release is widely deployed before we can do this?

**Reply by andrewleader** [reply]
Unfortunately this looks like in this case, it was a limitation in the OS itself. It's possible for us to try and service the fix down-level, but servicing fixes aren't ever guaranteed to make it down to older OS's (the user might have turned updates off completely, etc).

If this is a severe blocker for you (or anyone else), please let us know. Our hope is that in the future as we have more and more in Reunion, we'll be able to fix issues out-of-band, so this won't happen as often in the future.

**Reply by khdc** [reply]
This is a major blocker for my company. We create high-end engineering software and have utilities that need to run elevated. If WinUI 3 desktop can't run elevated, then why does it even exist?

**Reply by andrewleader** [reply]
Thank you @khdc for the feedback. We're aware this is a blocker for some companies, and adding elevation support is a high priority item on our list. We've captured your feedback.

**Reply by Sacwith** [reply]
Is there any update on adding elevation in winui3 app?

**Reply by KWodarczyk** [reply]
is there any workaround or is it possible to use pre release version if winui that has this feature implemented ?

**Reply by andrewleader** [reply]
@KWodarczyk elevation support is planned for 1.1, to subscribe to updates please vote on [the portal card](https://portal.productboard.com/winappsdk/1-windows-app-sdk/c/37-elevated-admin-support), and we'll be sure to let you know when it's available in pre-release form!

**Comment by driver1998** [comment]
How does Windows Terminal manage to do it if WinRT activation doesn't work in elevated apps? 🤔

**Reply by DrusTheAxe** [reply]
To clarify @andrewleader's comment, WinUI3 **in an unpackaged app** does not currently support elevation because the underlying tech to support unpackaged apps ([Dynamic Dependencies](https://github.com/microsoft/WindowsAppSDK/blob/main/specs/dynamicdependencies/DynamicDependencies.md)) does not currently support elevation. That's due to a limitation of Packaged COM.

This is an unpackaged app issue. Terminal is a packaged app so it doesn't need [Dynamic Dependencies](https://github.com/microsoft/WindowsAppSDK/blob/main/specs/dynamicdependencies/DynamicDependencies.md) to use WinUI.

This is a known issue tracked by [Dynamic Dependencies doesn't support Elevation #567](https://github.com/microsoft/WindowsAppSDK/issues/567). We're considering alternative (non-Packaged COM) designs to address #567. If this interests you please share your input at that issue to help us prioritize our efforts.

**Reply by andrewleader** [reply]
@DrusTheAxe actually WinUI 3 in **packaged apps** also does NOT support elevation. It's an OS-level issue, COM/WinRT doesn't allow per-user class registrations to be activated from an elevated process. Inbox APIs work when elevated, but WinAppSDK isn't inbox. We've fixed that in the OS in the next release.

@driver1998 I don't think Windows Terminal is using WinUI 3 or WinAppSDK yet? I believe they're still on WinUI 2.x.

**Reply by zadjii-msft** [reply]
That's correct, the Terminal is on WinUI 2.x.

**Reply by DrusTheAxe** [reply]
>WinUI 3 in packaged apps also does NOT support elevation. It's an OS-level issue, COM/WinRT doesn't allow

Doh! I thought the limitation was just Packaged COM, didn't realized packaged WinRT's also affected. Thanks for the correction.

@andrewleader given this limitation on downlevel systems we may want to consider adding an explicit check (if OS<N and IsElevated) to produce a clear error e.g. in XAML's generated main. We added an IsElevated/error check in Dynamic Dependencies because if you tried you'd get hard to understand errors due to being elevated (e.g. ERROR_NOT_FOUND).

**Reply by DHowett** [reply]
@andrewleader Terminal runs elevated by bundling all of its dependencies in its package and using 19H1's SXS WinRT Activation manifest support. :smile:

It does so whether it's inside or outside its package.

**Reply by ghost** [reply]
> given this limitation on downlevel systems we may want to consider adding an explicit check (if OS<N and IsElevated) to produce a clear error e.g. in XAML's generated main.

@DrusTheAxe  consider doing so. this is going to cause a  headache for someone on downlevel windows versions who will try to use newer WinRT apis like https://github.com/microsoft/WindowsAppSDK/pull/1094 , https://github.com/microsoft/WindowsAppSDK/pull/1064, https://github.com/microsoft/WindowsAppSDK/pull/682

**Comment by Sacwith** [comment]
Any updates on adding elevation(uiAccess=true) in winui3?

**Reply by andrewleader** [reply]
Hi @Sacwith, elevation support is tentatively planned for 1.1 and you can subscribe to updates on it here: https://portal.productboard.com/winappsdk/1-windows-app-sdk/c/37-elevated-admin-support

**Reply by Sacwith** [reply]
@andrewleader  can we set uiAcess=true in appmanifest in Winui3 application? we need to make overlay come top so.

**Reply by andrewleader** [reply]
@Sacwith I can check to see if the team knows whether `uiAccess=true` is supported, in the meantime have you tried using that?

**Reply by Sacwith** [reply]
@andrewleader Yes i have tried using that but some exception so.

**Reply by Sacwith** [reply]
@andrewleade
error:  the requested operation requires elevation

**Reply by andrewleader** [reply]
If the error says it requires elevation, elevation is not supported in 1.0 but is on our 1.1 plans. So this isn't supported, but should work in 1.1. Thanks!

**Reply by harishsk** [reply]
Hi, Any idea when we can expect support for running WinUI3 apps with elevated privilege? I get this error on Win11.

Unable to activate Windows Store app '<guid>!App'. The activation request failed with error 'The request is not supported'.

**Reply by DrusTheAxe** [reply]
> Hi, Any idea when we can expect support for running WinUI3 apps with elevated privilege? I get this error on Win11.

As @andrewleader noted it's coming in 1.1 (as noted [on the Roadmap](https://github.com/microsoft/WindowsAppSDK/#roadmap)).

**Comment by MaddalaN** [comment]
I had the same issue, WinUI3 app was not launching in elevated mode but worked fine in unelevated mode.

Visual Studio output window logs below message when started in admin mode
`
The program '[20848] TestApp.exe: Program Trace' has exited with code 0 (0x0).
The program '[20848] TestApp.exe' has exited with code 4294967295 (0xffffffff).
`
It would have been nice if it logs some message saying 'run unelevated' or something like that so that it saves some dev time. Thanks!!
